rfcs/0012-UI5-Tooling-Extension-API-3.md

```diff
@@ -0,0 +1,415 @@
+
+## Proposed Enhancements
+<!-- You can either remove the following explanatory text or move it into this comment for later reference -->
+
+### 1. Dependencies to UI5 Tooling Modules
+
+#### Description
+
+It is common for extensions to require UI5 Tooling modules like `@ui5/fs` or `@ui5/logger`. For example to create [`Resource`](https://sap.github.io/ui5-tooling/api/module-@ui5_fs.Resource.html) entities, or to log messages to the console.
+
+The problem that arises with this, is **compatibility across UI5 Tooling versions**.
+
+#### Examples
+
+1. A `Resource` created with **`@ui5/fs` v2 is not compatible to `@ui5/fs` v3** and needs to be migrated by UI5 Tooling v3.
+2. Extensions using **`@ui5/logger` v2 will use a different logger instance than UI5 Tooling v3**. This can lead to, for example, log-level configuration not being available to the extension.
+
+#### Solution
+
+**Dependency injection.** By providing extensions with callback functions to create resources or log messages, the extension can remove problematic dependencies like `@ui5/fs` or `@ui5/logger` from their package.json.
+
+In a future **UI5 Tooling version 4**, extensions defining **Specification Version 3** would then still receive the expected `Resource` and `Logger` APIs, thus staying compatible while UI5 Tooling v4 makes internal enhancements and changes to the APIs.
+
+The new APIs should be provided via the existing helper classes [`taskUtil`](https://sap.github.io/ui5-tooling/api/module-@ui5_builder.tasks.TaskUtil.html) and [`middlewareUtil`](https://sap.github.io/ui5-tooling/api/module-@ui5_server.middleware.MiddlewareUtil.html).
+
+#### New API
+* New parameter **log**
+    * The provided logger instance will automatically have the name of the custom task or middleware.
+    * Additional sub-loggers can be crated using `log.createSubLogger("name suffix")`
+* taskUtil|middlewareUtil.**resourceFactory.createResource(**_{...}_**)**
+    * Creates and returns a `Resource` with an interface signature according to the Specification Version of the extension.
+    * This implies a new `resourceFactory` object which provides Specification Version-dependent access to functions of the [`@ui5/fs/resourceFactory`](https://sap.github.io/ui5-tooling/api/module-@ui5_fs.resourceFactory.html)
+
+#### Solution Example
+
+Migrating the custom task example from our current [documentation](https://sap.github.io/ui5-tooling/pages/extensibility/CustomTasks/#example-libtasksgeneratemarkdownfilesjs) to make use of the new API could look like this:
+
+```patch
+const path = require("path");
+-const {Resource} = require("@ui5/fs");
+-const log = require("@ui5/logger").getLogger("builder:customtask:markdown");
+const renderMarkdown = require("./renderMarkdown");
+
+-module.exports = async function({workspace, dependencies, taskUtil, options}) {
++module.exports = async function({workspace, dependencies, log, taskUtil, options}) {
```

Comment:
should we use here an ESM export to promote the new feature?

Reply:
See https://github.com/SAP/ui5-tooling/pull/664#discussion_r1048258008